auth/simple_token.go

```diff
@@ -21,13 +21,104 @@ import (
 	"crypto/rand"
 	"math/big"
 	"strings"
+	"sync"
+	"time"
 )
 
 const (
 	letters                  = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
 	defaultSimpleTokenLength = 16
+	simpleTokenTTL           = 5 * time.Minute
+	simpleTokenTTLResolution = 1 * time.Second
 )
 
+type simpleTokenTTLKeeper struct {
+	tokens              map[string]time.Time
+	addSimpleTokenCh    chan string
+	resetSimpleTokenCh  chan string
+	deleteSimpleTokenCh chan string
+	stopCh              chan chan struct{}
+	deleteTokenFunc     func(string)
+	mu                  sync.Mutex
+	running             bool
+}
+
+func NewSimpleTokenTTLKeeper(deletefunc func(string)) *simpleTokenTTLKeeper {
+	stk := &simpleTokenTTLKeeper{
+		tokens:              make(map[string]time.Time),
+		addSimpleTokenCh:    make(chan string, 1),
+		resetSimpleTokenCh:  make(chan string, 1),
+		deleteSimpleTokenCh: make(chan string, 1),
+		stopCh:              make(chan chan struct{}),
+		deleteTokenFunc:     deletefunc,
+		mu:                  sync.Mutex{},
+	}
+	return stk
+}
+
+func (tm *simpleTokenTTLKeeper) start() {
+	go tm.run()
+}
+
+func (tm *simpleTokenTTLKeeper) stop() {
+	tm.mu.Lock()
+	defer tm.mu.Unlock()
+	if tm.running {
+		stoppedCh := make(chan struct{})
+		tm.stopCh <- stoppedCh
+		<-stoppedCh
+		close(stoppedCh)
+		tm.tokens = make(map[string]time.Time)
+	}
+}
+
+func (tm *simpleTokenTTLKeeper) addSimpleToken(token string) {
+	tm.addSimpleTokenCh <- token
+}
+
+func (tm *simpleTokenTTLKeeper) resetSimpleToken(token string) {
+	tm.resetSimpleTokenCh <- token
+}
+
+func (tm *simpleTokenTTLKeeper) deleteSimpleToken(token string) {
+	tm.deleteSimpleTokenCh <- token
+}
+func (tm *simpleTokenTTLKeeper) run() {
+	tokenTicker := time.NewTicker(simpleTokenTTLResolution)
+	defer tokenTicker.Stop()
+	tm.mu.Lock()
+	tm.running = true
+	tm.mu.Unlock()
+	plog.Infof("starting the simple token ttl keeper routine")
+	for {
+		select {
+		case t := <-tm.addSimpleTokenCh:
+			tm.tokens[t] = time.Now().Add(simpleTokenTTL)
+		case t := <-tm.resetSimpleTokenCh:
+			if _, ok := tm.tokens[t]; ok {
+				tm.tokens[t] = time.Now().Add(simpleTokenTTL)
+			}
+		case t := <-tm.deleteSimpleTokenCh:
+			delete(tm.tokens, t)
+		case <-tokenTicker.C:
+			nowtime := time.Now()
+			for t, tokenendtime := range tm.tokens {
+				if nowtime.After(tokenendtime) {
+					tm.deleteTokenFunc(t)
+					delete(tm.tokens, t)
+				}
+			}
+		case stopped := <-tm.stopCh:
+			stopped <- struct{}{}
+			tm.mu.Lock()
```

Comment:
this doesn't work

T1: call stop(), acquire lock, wait on <-stopped
T2: call stop(), wait on lock
T3: stopped <- struct{}{}
T1: gets <-stopped, release lock
T2: acquires lock, wait on <-stopped
T3: try to acquire lock to set running = false, but already acquired; deadlock.